PythonPractice/TicTacToe/tictactoe.py

```diff
@@ -0,0 +1,155 @@
+
+    def initialize_board(self):
+        """
+        Creates an empty board to be played on
+        """
+        # get board size from user
+        good_input = False
+        size = None
+        while not good_input:
+            try:
+                size = int(raw_input("What size board to do you want to play on?\n"))
+                if size < 3:
+                    raise BoardTooSmallError
+            except BoardTooSmallError:
+                print "Your board must be at least a 3x3 to play."
+            except ValueError:
+                print "Your board size must be an integer."
+            else:
+                good_input = True
+        # initialize board to empty strings
+        self.board = [[None for row in range(0, size)] for col in range(0, size)]
+        self.size = len(self.board)
+        self.num_turns = size*size
+        self.draw_board()
+
+    def draw_board(self):
+        """
+        Draws current state of the board
+        """
+        for row in self.board:
+            print "---" * self.size
+            for col in row:
+                if col:
+                    print "|%s" % col,
+                else:
+                    print "| ",
+            print "|"
+        print "---" * self.size
+
+    def play(self):
+        """
+        If there are turns left, allows a play
+        Makes sure the user input for row and column is valid (positive integer, untake space, in bounds)
+        Sets the space according to the player's turn
+        Switches player before exiting
+        Output:
+            True if there are remaining turns
+            False if there are no remaining turns
+        """
+        if self.num_turns > 0:
+            good_input = False
+            row = None
+            col = None
+            # communicate who should play next
+            print "-----------------------------"
+            print "PLAYER %s's turn" % self.player
+            # get validated user input
+            while not good_input:
+                try:
+                    row = int(raw_input("What row?\n"))-1
+                    col = int(raw_input("What column?\n"))-1
+                    if row < 0 or col < 0:
+                        raise ValueError
+                    if self.board[row][col]:
+                        raise SpaceTakenError
+                    if row > self.size or col > self.size:
+                        raise IndexError
+                except ValueError:
+                    print "Make sure the row and column are positive integers."
+                except SpaceTakenError:
+                    print "That space is already taken. Try again."
+                except IndexError:
+                    print "That space is out of bounds. Try again."
+                else:
+                    good_input = True
+            # set space with user's value (X or O) depending on whose turn it is
+            # update next player
+            if self.player == self.playerX:
```

Comment:
There may be little practical point in a tic-tac-toe game with more (or fewer) than two players, but by way of a thought experiment, how might you change the TicTacToe class such that turns cycled through an arbitrary number of players, each represented by a different letter, with the players and their letters set in the call to the constructor?